### PR TITLE
[Core][Pubsub] handle subscriber client-side connection issues with sequence numbers

### PR DESCRIPTION
**Pull request by mwtian**
## Why are these changes needed?
Currently when a publisher replies to a subscriber long poll request, there is no acknowledgement of subscriber receiving the data. Also publisher immediately deletes the replied messages, regardless of whether they actually are received by the subscriber. This PR adds a sequence number to the long poll request and response, so publisher sends not-yet-received messages to the subscriber, and deletes messages already processed at the subscriber. This allows re-enabling two CI tests, and helps supporting reconnection to GCS later.

This does not fix all cases where a message can be dropped. For example, the reply from publisher could be dropped, while the subscriber does not detect connection issue or make another polling request. Then the subscriber state with sequence number can time out at the publisher (e.g. after 30s). This can be addressed in future with client timeout / deadline on polling requests.

This affects existing Ray pubsub usages. It should probably merge at least after 1.9 branch cut.

An integration test exercising publisher and subscriber together will be added when reconnection tests are fixed. They can be added in this PR too if desired.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by ericl — COMMENTED** [review]
I have a question on the fault-tolerance model we are trying to implement. It seems that this PR is assuming the connection can be faulty. But what if we destroyed the subscriber and generated a new subscriber id on faults?

Then there is no need to handle faults within a connection; the subscriber can re-connect with a different ID and receive the same data again right?

This is assuming the state-notification model (i.e., get latest state only), and we are ok with losing some intermediate data such as logs in the case of failure.

**Comment by mwtian** [comment]
Chatted with @ericl off thread. Will talk to @iycheng next about the design here. This is not blocking work on Redis removal, but it helps addressing some failure cases which we have in the CI.

**Comment by iycheng** [comment]
I synced it with @mwtian today and we feel it's more convent to use seq given the current situation (no grpc streaming).

First of all, reconnect sometimes is not an easy job. For example, in cpp, detect error and ask for reconnection, sometimes it's tricky to do in app layer. First of all you need to get the return code in callback, and callback need to ask the pubsub for reconnection, but the callback can't capture the pubsub handler.

It's also more aligned with the current resource broadcasting model. Right now, we use seq to detect message losing, it'll be easier to convert to that later.

Another thing is that when the connection is broken, the app usually just reconnect and gets whatever message the app can get and the app still needs a seq or something to know what's the thing got lost if the app can recover. So I'll say, it's not really saving a lot for this case.

One issue for seq is that users might forget to check it. But we think it's ok since it's an internal component, we'll just need to use it correctly.

File: src/ray/protobuf/pubsub.proto [hunk 1]
```diff
@@ -186,11 +186,18 @@ message WorkerObjectLocationsSubMessage {
...
 }
 
 message PubsubLongPollingReply {
   /// The messages that are published.
   repeated PubMessage pub_messages = 1;
+  /// Sequence number of this polling reply.
+  int64 seq = 2;
+  /// Whether the subscriber should reset its sequence number based on this reply,
+  /// e.g. when the subscriber makes an initial connection, or is likely re-connecting.
+  bool reset_seq = 3;
```
Comment (iycheng) [inline src/ray/protobuf/pubsub.proto:200]: is reset_seq a good name? should it be something like is_new_connection?

File: src/ray/pubsub/publisher.cc [hunk 2]
```diff
@@ -175,18 +184,28 @@ void Subscriber::QueueMessage(const rpc::PubMessage &pub_message, bool try_publi
...
 
-  if (!force_noop) {
-    // Reply to the long polling subscriber. Swap the reply here to avoid extra copy.
-    long_polling_connection_->reply->Swap(mailbox_.front().get());
-    mailbox_.pop();
+  // No message should have been added to the reply.
+  RAY_CHECK(long_polling_connection_->reply->pub_messages().empty());
+  if (force_noop) {
+    // Use -1 sequence number to mark a no-op message.
+    long_polling_connection_->reply->set_seq(-1);
```
Comment (iycheng) [inline src/ray/pubsub/publisher.cc:199]: why set -1 here? shouldn't we just always increase this for each message?

**Comment by rkooo567** [comment]
I will leave review tomorrow. Btw, since it changes the existing pubsub module's behavior, and this module is on sensitive path (reference counting + object eviction + obod), can we try running nightly tests 2~3 times before merging it?

**Comment by rkooo567** [comment]
Btw, can we update the pubusb README.md or write some documentation regarding how the behavior will change?

**Comment by jovany-wang** [comment]
+1 for using seqno. State-based model requires that we should cover businesses in many places which is easy to be missed. seqno here can make our fault handling here more cohesive.

**Comment by jovany-wang** [comment]
>Currently when a publisher replies to a subscriber long poll request, there is no acknowledgement of subscriber receiving the data. Also publisher immediately deletes the replied messages, regardless of whether they actually are received by the subscriber. This PR adds a sequence number to the long poll request and response, so publisher sends not-yet-received messages to the subscriber, and deletes messages already processed at the subscriber.

I can understand this case you described, but I have no any context on the case in our codebase, could you provide more information? For example, which test failed.

**Pull request closed without merging**